generator/template/Emoji.java

```diff
@@ -44,27 +62,52 @@
   }
 
   @NonNull @Override public Drawable getDrawable(final Context context) {
-    if (sheet == null) {
-      synchronized (LOCK) {
-        if (sheet == null) {
-          sheet = BitmapFactory.decodeResource(context.getResources(), R.drawable.emoji_<%= package %>_sheet);
-        }
-      }
+    final CacheKey key = new CacheKey(x, y);
+    final Bitmap bitmap = BITMAP_CACHE.get(key);
+    if (bitmap != null) {
+      return new BitmapDrawable(context.getResources(), bitmap);
     }
-
-    final Bitmap cut = Bitmap.createBitmap(sheet, x * 66, y * 66, 64, 64);
-
+    final Bitmap strip = loadStrip(context);
+    final Bitmap cut = Bitmap.createBitmap(strip, 1,
+        y * SPRITE_SIZE_INC_BORDER + 1, SPRITE_SIZE, SPRITE_SIZE);
+    BITMAP_CACHE.put(key, cut);
```

Comment:
I added the LRU cache because otherwise it was possible to trigger a lot of loading and unloading of strips when the app was under memory pressure. For example, I'd be looking at a conversation with a few emoji visible, the GC would run and unload the strips backing the emoji, then the next time the conversation was redrawn, the strips would be reloaded. I didn't do any rigorous performance testing, but my guess was that this repeated loading was probably worse for performance overall than keeping a few recently used sprites in memory.

If you'd like to do a proper performance comparison, the version without the LRU cache is here: https://github.com/akwizgran/Emoji/tree/sprite-strips-soft-refs

Reply:
I see! I'll compare both solutions.